File: artifacts/tulipindicators/decompose/src/willr.c
int ti_willr_start(const double *options)
{
  return ((int) options[0]) - 1;
}


/*** DEPENDENCIES:
***/


// hint:  ['output_idx_ref is a mutable refrence to unsigned int', 'maxi_ref is a mutable refrence to int', 'mini_ref is a mutable refrence to int', 'max_ref is a mutable refrence to double', 'min_ref is a mutable refrence to double', 'bar_ref is a mutable refrence to double', 'j_ref is a mutable refrence to int']
void helper_ti_willr_1(unsigned int * const output_idx_ref, int * const maxi_ref, int * const mini_ref, double * const max_ref, double * const min_ref, double * const bar_ref, int * const j_ref, const double * const high, const double * const low, const double * const close, double * const output, int trail, int i)
{
  unsigned int output_idx = *output_idx_ref;
  int maxi = *maxi_ref;
  int mini = *mini_ref;
  double max = *max_ref;
  double min = *min_ref;
  double bar = *bar_ref;
  int j = *j_ref;
  bar = high[i];
  if (maxi < trail)
  {
    maxi = trail;
    max = high[maxi];
    j = trail;
    while ((++j) <= i)
    {
      bar = high[j];
      if (bar >= max)
      {
        max = bar;
        maxi = j;
      }
    }

  }
  else
    if (bar >= max)
  {
    maxi = i;
    max = bar;
  }
  bar = low[i];
  if (mini < trail)
  {
    mini = trail;
    min = low[mini];
    j = trail;
    while ((++j) <= i)
    {
      bar = low[j];
      if (bar <= min)
      {
        min = bar;
        mini = j;
      }
    }

  }
  else
    if (bar <= min)
  {
    mini = i;
    min = bar;
  }
  const double highlow = max - min;
  const double r = (highlow == 0.0) ? (0.0) : ((-100) * ((max - close[i]) / highlow));
  output[output_idx] = r;
  output_idx += 1;
  *output_idx_ref = output_idx;
  *maxi_ref = maxi;
  *mini_ref = mini;
  *max_ref = max;
  *min_ref = min;
  *bar_ref = bar;
  *j_ref = j;
}


/*** DEPENDENCIES:
***/


int ti_willr(int size, const double * const *inputs, const double *options, double * const *outputs)
{
  const double *high = inputs[0];
  const double *low = inputs[1];
  const double *close = inputs[2];
  const int period = (int) options[0];
  double *output = outputs[0];
  unsigned int output_idx = 0;
  if (period < 1)
  {
    return 1;
  }
  if (size <= ti_willr_start(options))
  {
    return 0;
  }
  int trail = 0;
  int maxi = -1;
  int mini = -1;
  double max = high[0];
  double min = low[0];
  double bar;
  int i;
  int j;
  for (i = period - 1; i < size; i += 1, trail += 1)
  {
    helper_ti_willr_1(&output_idx, &maxi, &mini, &max, &min, &bar, &j, high, low, close, output, trail, i);
  }

  assert(((&output[output_idx]) - outputs[0]) == (size - ti_willr_start(options)));
  return 0;
}


/*** DEPENDENCIES:
void helper_ti_willr_1(unsigned int * const output_idx_ref, int * const maxi_ref, int * const mini_ref, double * const max_ref, double * const min_ref, double * const bar_ref, int * const j_ref, const double * const high, const double * const low, const double * const close, double * const output, int trail, int i)
{
  unsigned int output_idx = *output_idx_ref;
  int maxi = *maxi_ref;
  int mini = *mini_ref;
  double max = *max_ref;
  double min = *min_ref;
  double bar = *bar_ref;
  int j = *j_ref;
  bar = high[i];
  if (maxi < trail)
  {
    maxi = trail;
    max = high[maxi];
    j = trail;
    while ((++j) <= i)
    {
      bar = high[j];
      if (bar >= max)
      {
        max = bar;
        maxi = j;
      }
    }

  }
  else
    if (bar >= max)
  {
    maxi = i;
    max = bar;
  }
  bar = low[i];
  if (mini < trail)
  {
    mini = trail;
    min = low[mini];
    j = trail;
    while ((++j) <= i)
    {
      bar = low[j];
      if (bar <= min)
      {
        min = bar;
        mini = j;
      }
    }

  }
  else
    if (bar <= min)
  {
    mini = i;
    min = bar;
  }
  const double highlow = max - min;
  const double r = (highlow == 0.0) ? (0.0) : ((-100) * ((max - close[i]) / highlow));
  output[output_idx] = r;
  output_idx += 1;
  *output_idx_ref = output_idx;
  *maxi_ref = maxi;
  *mini_ref = mini;
  *max_ref = max;
  *min_ref = min;
  *bar_ref = bar;
  *j_ref = j;
}


----------------------------
int ti_willr_start(const double *options)
{
  return ((int) options[0]) - 1;
}


----------------------------
***/

File: artifacts/tulipindicators/translate_ablation/src/willr.rs
use crate::*;
use rand::Rng;

pub fn ti_willr_start(options: Option<&[f64]>) -> Option<i32> {
    let options = options?;
    if options.is_empty() {
        return None;
    }
    Some(options[0] as i32 - 1)
}

pub fn ti_willr(size: i32, inputs: &[&[f64]], options: &[f64], outputs: &mut [&mut [f64]]) -> i32 {
    let high = inputs[0];
    let low = inputs[1];
    let close = inputs[2];
    let period = options[0] as i32;
    let output = &mut outputs[0];
    
    if period < 1 {
        return 1;
    }
    
    let start = match ti_willr_start(Some(options)) {
        Some(s) => s,
        None => return 1,
    };
    
    if size <= start {
        return 0;
    }
    
    let mut trail = 0;
    let mut maxi = -1;
    let mut mini = -1;
    let mut max_val = high[0];
    let mut min_val = low[0];
    let mut bar;
    
    let mut output_idx = 0;
    
    for i in (period - 1) as usize..size as usize {
        bar = high[i];
        if maxi < trail {
            maxi = trail;
            max_val = high[maxi as usize];
            let mut j = trail;
            while j <= i as i32 {
                bar = high[j as usize];
                if bar >= max_val {
                    max_val = bar;
                    maxi = j;
                }
                j += 1;
            }
        } else if bar >= max_val {
            maxi = i as i32;
            max_val = bar;
        }
        
        bar = low[i];
        if mini < trail {
            mini = trail;
            min_val = low[mini as usize];
            let mut j = trail;
            while j <= i as i32 {
                bar = low[j as usize];
                if bar <= min_val {
                    min_val = bar;
                    mini = j;
                }
                j += 1;
            }
        } else if bar <= min_val {
            mini = i as i32;
            min_val = bar;
        }
        
        let highlow = max_val - min_val;
        let r = if highlow == 0.0 {
            0.0
        } else {
            -100.0 * ((max_val - close[i]) / highlow)
        };
        
        output[output_idx] = r;
        output_idx += 1;
        trail += 1;
    }
    
    assert!((output_idx as i32) == (size - start));
    0
}
